x/mongo/driver/topology/pool.go

```diff
@@ -305,41 +205,105 @@ func (p *pool) makeNewConnection() (*connection, string, error) {
 				Reason:       event.ReasonPoolClosed,
 			})
 		}
-		_ = p.closeConnection(c) // The pool is disconnected or disconnecting, ignore the error from closing the connection.
-		return nil, event.ReasonPoolClosed, ErrPoolDisconnected
+		p.closeConnection(c) // The pool is disconnected or disconnecting, ignore the error from closing the connection.
+		return ErrPoolDisconnected
 	}
 
 	p.Lock()
 	p.opened[c.poolID] = c
 	p.Unlock()
 
-	return c, "", nil
+	go p.connectAndAdd(c)
 
+	return nil
 }
 
-func (p *pool) pinConnectionToCursor() {
-	atomic.AddUint64(&p.pinnedCursorConnections, 1)
+// connectAndAdd connects and then adds the connection to the buffered channel
+func (p *pool) connectAndAdd(c *connection) {
+	start := time.Now()
+	c.connect(context.Background())
+	err := c.wait()
+	if err != nil {
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionClosed,
+				Address:      p.address.String(),
+				ConnectionID: c.poolID,
+				Reason:       event.ReasonTimedOut,
+			})
+		}
+
+		p.Lock()
+		delete(p.opened, c.poolID)
+		p.Unlock()
+		return
+	}
+
+	if p.monitor != nil {
+		elapsed := time.Since(start)
+		p.monitor.Event(&event.PoolEvent{
+			Type:         event.ConnectionConnected,
+			Address:      p.address.String(),
+			ConnectionID: c.poolID,
+			Duration:     &elapsed,
+		})
+	}
+
+	p.poolChan <- c
 }
 
-func (p *pool) unpinConnectionFromCursor() {
-	// See https://golang.org/pkg/sync/atomic/#AddUint64 for an explanation of the ^uint64(0) syntax.
-	atomic.AddUint64(&p.pinnedCursorConnections, ^uint64(0))
+// manager spins up connections or reduces connections as needed
+func (p *pool) manager(doneChan chan struct{}) {
+	ticker := time.NewTicker(maintainInterval)
+	lastCloseTime := time.Now()
+	for {
+		select {
+		case <-doneChan:
+			return
+		case <-ticker.C:
+			if p.connectionNeeded() {
+				p.addConnection()
```

Comment:
The manager routine runs at every `maintainInterval`. Since this creates one connection at a time, this may have a caveat when multiple new connections are created immediately.

For example, if a caller configures a `MinPoolSize=0` and 10 concurrent requests come in for new connections, this may effectively offset those connections at 10ms increments.

Reply:
Correct, there is a maximum ramping up rate with this implementation.